engine/classes/Elgg/EntityIconService.php

```diff
@@ -0,0 +1,404 @@
+		$tmp->close();
+		copy($filename, $tmp->getFilenameOnFilestore());
+
+		$tmp->mimetype = (new MimeTypeDetector())->getType($tmp->getFilenameOnFilestore());
+		$tmp->simpletype = elgg_get_file_simple_type($tmp->mimetype);
+
+		$result = $this->saveIcon($entity, $tmp, $type, $coords, false);
+
+		$tmp->delete();
+
+		return $result;
+	}
+
+	/**
+	 * Saves icons using a file located in the data store as the source.
+	 *
+	 * @param ElggEntity $entity Entity to own the icons
+	 * @param string     $file   An ElggFile instance
+	 * @param string     $type   The name of the icon. e.g., 'icon', 'cover_photo'
+	 * @param array      $coords An array of cropping coordinates x1, y1, x2, y2
+	 * @return bool
+	 * @throws InvalidParameterException
+	 */
+	public function saveIconFromElggFile(ElggEntity $entity, ElggFile $file, $type = 'icon', array $coords = array()) {
+		if (!$file->exists()) {
+			throw new InvalidParameterException(__METHOD__ . ' expects an instance of ElggFile with an existing file on filestore');
+		}
+
+		$tmp_filename = time() . pathinfo($file->getFilenameOnFilestore(), PATHINFO_BASENAME);
+		$tmp = new ElggFile();
+		$tmp->owner_guid = $entity->guid;
+		$tmp->setFilename("tmp/$tmp_filename");
+		$tmp->open('write');
+		$tmp->close();
+		copy($file->getFilenameOnFilestore(), $tmp->getFilenameOnFilestore());
+
+		$tmp->mimetype = (new MimeTypeDetector())->getType($tmp->getFilenameOnFilestore(), $file->getMimeType());
+		$tmp->simpletype = elgg_get_file_simple_type($tmp->mimetype);
+
+		$result = $this->saveIcon($entity, $tmp, $type, $coords, false);
+
+		$tmp->delete();
+
+		return $result;
+	}
+
+	/**
+	 * Saves icons using a created temporary file
+	 *
+	 * @param ElggEntity $entity        Temporary ElggFile instance
+	 * @param string     $file          Temporary ElggFile instance
+	 * @param string     $type          The name of the icon. e.g., 'icon', 'cover_photo'
+	 * @param array      $coords        An array of cropping coordinates x1, y1, x2, y2
+	 * @param bool       $keep_uploaded Keep uploaded file under /icons/$type/_uploaded.jpg
+	 * @return bool
+	 */
+	public function saveIcon(ElggEntity $entity, ElggFile $file, $type = 'icon', array $coords = array(), $keep_uploaded = true) {
+
+		// @todo: trigger a hook to allow plugins to manipulate source file
+		// e.g. extract video frame from uploaded video or rotate the image
+
+		// @todo: validate file type - should we bail if file is not image?
+		if (!$file->exists()) {
+			$this->logger->error('Source file passed to ' . __METHOD__ . ' can not be resolved to a valid image');
+			return false;
+		}
+
+		$entity_type = $entity->getType();
+		$entity_subtype = $entity->getSubtype();
+
+		$x1 = (int) elgg_extract('x1', $coords);
+		$y1 = (int) elgg_extract('y1', $coords);
+		$x2 = (int) elgg_extract('x2', $coords);
+		$y2 = (int) elgg_extract('y2', $coords);
+
+		$this->deleteIcon($entity, $type);
+
+		$success = function() use ($entity, $type, $x1, $y1, $x2, $y2) {
```

Comment:
Will be duplicate once I add the hooks
On Apr 11, 2016 4:25 PM, "Steve Clay" <notifications@github.com> wrote:

> In engine/classes/Elgg/EntityIconService.php
> <https://github.com/Elgg/Elgg/pull/9655#discussion_r59212678>:
>
> > +		if (!$file->exists()) {
> > +			$this->logger->error('Source file passed to ' . __METHOD__ . ' can not be resolved to a valid image');
> > +			return false;
> > +		}
> > +
> > +		$entity_type = $entity->getType();
> > +		$entity_subtype = $entity->getSubtype();
> > +
> > +		$x1 = (int) elgg_extract('x1', $coords);
> > +		$y1 = (int) elgg_extract('y1', $coords);
> > +		$x2 = (int) elgg_extract('x2', $coords);
> > +		$y2 = (int) elgg_extract('y2', $coords);
> > +
> > +		$this->deleteIcon($entity, $type);
> > +
> > +		$success = function() use ($entity, $type, $x1, $y1, $x2, $y2) {
>
> I seem them used only once, so just put the logic inline?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/Elgg/Elgg/pull/9655/files/cad4a4c171fb0d7cc6ae0b4454d7e9a007373fbb#r59212678>
>